### PR TITLE
WT-8973 Define semantics of zero timestamp in our APIs

### PR DESCRIPTION
**Pull request by keithbostic**

### CODE REVIEW

**Comment by keithbostic** [comment]
Reviewers -- I started with @agorrod's list from the ticket. Here's my final list and resolution.

WT_CONNECTION::query_timestamp() /* Specifically regards to whether 0 and not set need to be distinguishable */
WT_SESSION::query_timestamp() /* Specifically regards to whether 0 and not set need to be distinguishable */
	Changed; query_timestamp now returns a timestamp of 0 in all cases where the timestamp isn't set.

WT_SESSION::verify(cfg="stable_timestamp") /* As to whether setting this to 0 is different to not including the option */
WT_SESSION::create(cfg="import=[compare_timestamp]") /* As to whether setting this to 0 is different to not including the option */
	Ignored; these APIs don't have to change as they are timestamp configuration arguments, not actual timestamps.

WT_SESSION::flush_timestamp() /* As to whether setting this to 0 is different to not including the option */
	Ignored; the API is no longer needed and I removed it in this ticket.

/* As to whether setting this to 0 is different to not including the option */
WT_CONNECTION::set_timestamp()
WT_SESSION::set_timestamp()
WT_SESSION::set_timestamp_uint()
/* The following three functions set per-transaction timestamps. Check to see if there is a difference between not setting a timestamp, and explicitly setting it to 0 */
WT_SESSION::commit_transaction(cfg="commit_timestamp=X,durable_timestamp=X")
WT_SESSION::prepare_transaction()
WT_SESSION::timestamp_transaction()
        Changed; set_timestamp in all of its forms now rejects any attempt to set a timestamp of 0.

**Review by korteland — APPROVED** [review]
Looks good to me - thanks for the doc updates.

File: src/include/wiredtiger.in [hunk 1]
```diff
@@ -1893,18 +1890,19 @@ struct __wt_session {
...
-	 * timestamp_transaction when the string parsing done in that method becomes a bottleneck.
+	 * Set a timestamp on a transaction numerically.  Prefer this method over
+	 * WT_SESSION::timestamp_transaction if the hexadecimal string parsing done in that method
+	 * becomes a bottleneck.
+	 *
+	 * The WT_SESSION.timestamp_transaction_uint method can only be used at snapshot isolation.
 	 *
-	 * The WT_SESSION.timestamp_transaction_uint method can only be used at snapshot
-	 * isolation.
+	 * @snippet ex_all.c transaction timestamp_uint
```
[the remark this answers is not in the document]
Comment (agorrod) [reply src/include/wiredtiger.in:1899]: ex_all.c is used to generate the code snippets in our documentation. We pull them out of a functioning application to ensure our documentation includes code that will at least compile.

[the remark this answers is not in the document]
Comment (korteland) [reply src/include/wiredtiger.in:1899]: Cool, thanks makes sense. Thanks

**Review by agorrod — APPROVED** [review]
Thanks @keithbostic lgtm - like other timestamp API changes, I think this deserves a MongoDB patch build before merging into develop, but if that testing looks OK, I'm good to go with this.

File: dist/api_data.py [hunk 2]
```diff
@@ -1647,11 +1647,11 @@ def __ge__(self, other):
...
-        specify which timestamp to query: \c commit returns the most recently
-        set commit_timestamp; \c first_commit returns the first set
-        commit_timestamp; \c prepare returns the timestamp used in preparing a
-        transaction; \c read returns the timestamp at which the transaction is
-        reading at. See @ref timestamp_txn_api''',
+        specify which timestamp to query: \c commit returns the most
+        recently set commit_timestamp; \c first_commit returns the first set
+        commit_timestamp; \c prepare returns the timestamp used in preparing
+        a transaction; \c read returns the timestamp at which the transaction
+        is reading at. See @ref timestamp_txn_api''',
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply dist/api_data.py]: Fixed in 235d572.

[the remark this answers is not in the document]
Comment (korteland) [reply dist/api_data.py]: thanks!

**Review by kommiharibabu — APPROVED** [review]
LGTM

**Pull request merged**